[FEATURE]: Show image context menus on Mobile

Rikiub opened the pull request:
## Old behaviour

Disrupt the workflow and it's not saved in Android clipboard, so only I can share the image.

https://github.com/user-attachments/assets/b6886693-1a9e-475b-9e43-5d971e3c22e4

## New behavior

I can open a context menu with many options and copy images to clipboard more easily.

Also you can open the image in new tab, show in the file explorer or rename the file.

https://github.com/user-attachments/assets/3d9db56e-7f04-4efb-ae64-10163d14088c

## Extras

- Currently only tested on Android and Obsidian Desktop.
- Added localization for "copy to clipboard" notifications.
- Fixed #48 
- Added #54 so this closes #53

Rikiub committed:
Feature to rename image files from context menu

Rikiub committed:
Feature to rename image files from context menu

Rikiub committed:
Remove deno.lock

Rikiub committed:
Improve image copy on mobile

Rikiub committed:
Don't activate copy on canvases

Rikiub committed:
Show image context menus on Obsidian Mobile

- Show image context menus in any image
- Show image filename (only mobile)
- Simplied image copy to clipboard

Rikiub committed:
Merge mobile context menus and rename option

Rikiub committed:
Add rename feature

Rikiub committed:
urlPath.slice -> urlPath.substring

Just to keep consistency with main branch

Rikiub commented on src/main.ts:
```diff
@@ -55,104 +55,44 @@ export default class CopyUrlInPreview extends Plugin {
-            ];
-        } else {
-            offs = [
-                onElementToOff(document, "touchstart", "img",
```
Removed to avoid open the menu two times on mobile.
With "contextmenu" is enough.

Also this fixes #48

Rikiub committed:
Add localization to 'copy to clipboard' Notices

Rikiub committed:
Update README.md to reflect changes

Rikiub committed:
Little formatting change

Rikiub committed:
Open in new tab with TFile instead of image element

NomarCub committed:
npm audit fix

NomarCub committed:
add lint:fix command

NomarCub committed:
update linting dependencies

NomarCub committed:
bump version to 1.11

Rikiub committed:
Avoid zoom the image when context menu is opened on mobile

Rikiub committed:
Revert manifest.json to main

NomarCub committed:
remove unused type

Rikiub committed:
Use preventDefault instead of stopImmediatePropagation to get vibrations on mobile

NomarCub commented:
Hi!
I just started doing some modifications before your most recent push. Are you done with the stuff you wanted?
Also, what OS are you on? I tried your modifications, and the change in `getRelativePath` is off, my internal images get assigned as not internal, and only have the copy button and none of the others.

Rikiub commented:
I'm testing this on Linux and Android, what OS are you using?

Also It's weird, your images are detected as external images.

Can you give me a example? I mean, the markdown and the image that are you using.

NomarCub committed:
Merge branch 'feature/mobile-context-menu' of https://github.com/rikiub/obsidian-copy-url-in-preview into feature/mobile-context-menu

NomarCub committed:
npm run lint:fix

NomarCub commented:
I'm on Windows and in the case of `if (urlPath.startsWith(basePath)) {`, `urlPath` is `/C:/Users/...` and `basePath` is `C:\Users\...`. My implementation was messy, but this new one just doesn't work. If you have a nicer solution in mind I'd be glad to accept it, but otherwise just use the previous one, if it also works on Linux and Android.
Also, please fix the remaining lint issue.

Rikiub commented:
Yes, I can reproduce it
Let me try something

Rikiub committed:
Fix internal link detection on Windows

Rikiub commented:
Fixed and ready to merge!

And sorry but I don't know about the lint issue (I can't run eslint).

Rikiub committed:
Fixes recommended by deno linter

Rikiub committed:
Refactorization Part 1

Rikiub committed:
Refactorization Part 2

Rikiub committed:
Refactorization Part 2

Rikiub commented:
Maybe this is out of topic but...

Taking into account the large number of changes, I took the freedom to refact the code base. Many of my changes leave some functions useless, so I took the trouble of remove them.

NomarCub committed:
npm run lint:fix

NomarCub committed:
format

NomarCub committed:
minor refactors

NomarCub committed:
refactors and formatting

NomarCub commented:
Thanks for the feature, and the tidy up too!
I will examine it in more detail and test it myself, as the code is a bit brittle, and I don't use all the functionality myself. I'll make a release later this week or next week, and maybe try setting up a GitHub action for it beforehand too.

@Mara-Li would you be up for testing if the canvas functionality doesn't break with these changes?

Rikiub commented:
I use a lot of canvas so I already tested it.
I don't think should be issues but any thing tell me 👏

NomarCub commented:
Another thing: the file downloading workaround was for online images. They haven't worked properly in the plugin for a while, so I don't mind letting the downloading go, but I might need to reintroduce it later. How would you go about that one, reintroducing copying from external images on Android?

Rikiub commented:
Sorry, I don't understand.

With my changes copying from internal and external images still works (and the workaround is less complex). Also you can share the image on Android if you want.

Are you talking about a specific function that I removed? I would like if you link to the commit with this functionality to understand better.

Rikiub commented on src/helpers.ts:
```diff
@@ -23,16 +21,16 @@ export async function copyImageToClipboard(url: string | ArrayBuffer): Promise<v
 // https://developer.mozilla.org/en-US/docs/Web/HTML/CORS_enabled_image
 // option?: https://www.npmjs.com/package/html-to-image
-export async function loadImageBlob(imgSrc: string): Promise<Blob | null> {
+export function loadImageBlob(imgSrc: string): Promise<Blob | null> {
```
How I see the image downloading functionality is handled by `loadImageBlob` and used by `copyImageToClipboard`.

I don't know how handle `Blob` objects but I guess you could use `loadImageBlob` to get the `Blob` object and then convert it to a image.

NomarCub replied:
Yeah, copying online images is broken currently both on desktop and mobile. The mobile funytinality for it had the temporary file saving, that got removed now. I was just wondering if we'll need the temporary image file save later.

NomarCub replied:
But it'll probably work fine without it. Maybe a separate download menu item would be nice for mobile, or even desktop, but I definitely won't bring back the previous temporary save file mechanism, now that I think about it.

Rikiub replied:
Personally I think with `Copy to clipboard` is enough. The copy and share is handled by the system so it's no needed keep a temporal file.

The only use case I think is a feature to download and replace the external image with a internal image, but Obsidian already have a command to do this.

Rikiub committed:
Fixed: canvas context menu don't working on images

Rikiub committed:
Improve URl detection on canvas + Refactorization

Rikiub commented:
A suggestion, what about remove this setting?

![Screenshot_20250709-000434_Obsidian](https://github.com/user-attachments/assets/304ac5f8-2533-4f36-b74f-850bee0d19be)

The canvas menu already have all the options and activating this you will only get the image menu, so you are losing all the canvas menu funcionality. I don't see a use case for this.

NomarCub commented:
Please revert your latest change, or at least undo all the formatting changes and keep only the functional ones. I have to be able to keep maintaining the plugin and I don't need any more changes in this PR to keep track of.

As for your regular context menu option: also in a later PR.

Something broke the context menu in new windows on this branch, fix that, and only that, and I'll make a new release.

Rikiub committed:
Revert "Improve URl detection on canvas + Refactorization"

This reverts commit 2b3e258203fa5de35aed3ac976a834673cd6563d.

Rikiub committed:
Improve URl detection on canvas

Rikiub committed:
Fix context menu on new windows

Rikiub commented:
Sorry, just is hard read the code with the current formatting.

I reverted the formatting and fixed the context menu on new windows, should be enough.

NomarCub committed:
format

NomarCub committed:
add auto-release GitHub action from https://docs.obsidian.md/Plugins/Releasing/Release+your+plugin+with+GitHub+Actions

NomarCub commented:
Thank you for your patience and going along with my requests!
I really appreciate PRs like this that not just add stuff, but improve overall code quality and simplify things.
I'll merge now, and make a new release today. Feel free to send new PRs with improvements, although I might get cranky about formatting 😅.

NomarCub committed:
add Rikiub% to credits
